### PR TITLE
WIP: Allow to skip or whitelist methods client-gen generates

### PR DESCRIPTION
**Pull request by mfojtik**
This patch allows to choose what methods the client-gen should generate for specific resource.
The syntax is:

```
// +skipMethods=get
// +onlyMethods=create,delete
```

The first will skip generating `.Get` for the given resource. The second skips all methods except `.Create` and `.Delete`.

The reason for this patch is that some of our resources does not implement all CRUD methods that client-gen generates, but we still want clients to be available for them to use.

### CODE REVIEW

**Comment by mfojtik** [comment]
(need to update the fake generator as well)

**Comment by sttts** [comment]
Aren't the methods called verbs?

**Comment by sttts** [comment]
And something like `// +genclient:onlyVerbs=create,delete` would be clearer. We have similar `:` syntax elsewhere already. Output of the client context, the tags do not make much sense.

**Comment by mfojtik** [comment]
@sttts updating that syntax will be bigger change to the parser ;-) also inside the generator they call it 'methods' ;-)

**Comment by mfojtik** [comment]
@deads2k @sttts i agree that unifying the tag syntax in kube makes sense and should be done and you can sign me up to that in 1.8, but for now we need a way to generate these clients...

**Comment by deads2k** [comment]
> @sttts updating that syntax will be bigger change to the parser ;-) also inside the generator they call it 'methods' ;-)

The tag should probably be based on the user-facing names, since we'd want the freedom to refactor the clientset generator.



> Output of the client context, the tags do not make much sense.

You would have the lister/watcher and informer base itself on the client tags?  I could see that I guess, but knowing we'd want to use the power in three generators makes it seem more generic.

@mfojtik Whatever we decide about the tag format is going to have to be in the first iteration of this.  Otherwise we're just producing debt for the next rebaser to try to unwind.  We should at least try.

**Comment by mfojtik** [comment]
@deads2k ok i'll try :)

**Comment by sttts** [comment]
Don't have a strong opinion about the tag names. But, I would avoid introducing more complex syntax which subsumes multiple tags in one, i.e. **no** `+genclient:nonNamespaced=true,onlyMethods=foo` or anything similar, but `+nonNamespaced=true` plus `+onlyMethods=foo`, i.e. key=val syntax.

**Comment by mfojtik** [comment]
@sttts 👍 

```
// +genclient
// +genclient:nonNamespaced
// +genclient:onlyVerbs=create,delete
// +genclient:skipVerbs=get,update
```

wdyt?

**Comment by mfojtik** [comment]
dunno if we also want to namespace this with `+k8s:` the only reason i see is if we will in future wrap the upstream client-gen and provide custom options?

**Comment by sttts** [comment]
Looks good. this left out `=true` is optional?

**Comment by mfojtik** [comment]
@sttts do we need explicit bool? if you don't want the client to be generated just don't use the tag ;-)

**Comment by sttts** [comment]
for some generators we use `=false` to exclude types if the generation is enabled on the package level with `=package`, e.g. deepcopy. So `=true` is natural.

**Comment by mfojtik** [comment]
@sttts updated... check if you like this version and I will start upstream work.

File: pkg/authorization/generated/clientset/typed/authorization/v1/authorization_client.go [hunk 1]
```diff
@@ -40,20 +40,20 @@ func (c *AuthorizationV1Client) ClusterRoleBindings() ClusterRoleBindingInterfac
...
 
-func (c *AuthorizationV1Client) PolicyBindings(namespace string) PolicyBindingInterface {
-	return newPolicyBindings(c, namespace)
+func (c *AuthorizationV1Client) PolicyBindings() PolicyBindingInterface {
+	return newPolicyBindings(c)
 }
 
-func (c *AuthorizationV1Client) Roles(namespace string) RoleInterface {
-	return newRoles(c, namespace)
+func (c *AuthorizationV1Client) Roles() RoleInterface {
```
Comment (mfojtik) [inline pkg/authorization/generated/clientset/typed/authorization/v1/authorization_client.go]: @deads2k @sttts are these supposed to be namespaced or non-namespaced? not sure why this change, it might be bug in my code OR bug in existing code ;-)

File: pkg/image/generated/clientset/typed/image/v1/fake/fake_image_client.go [hunk 2]
```diff
@@ -14,8 +14,20 @@ func (c *FakeImageV1) Images() v1.ImageResourceInterface {
 	return &FakeImages{c}
 }
 
-func (c *FakeImageV1) ImageStreams(namespace string) v1.ImageStreamInterface {
-	return &FakeImageStreams{c, namespace}
+func (c *FakeImageV1) ImageSignatures() v1.ImageSignatureInterface {
+	return &FakeImageSignatures{c}
+}
+
+func (c *FakeImageV1) ImageStreams() v1.ImageStreamInterface {
```
Comment (mfojtik) [inline pkg/image/generated/clientset/typed/image/v1/fake/fake_image_client.go]: ok imagestreams are definitely namespaced, so bug in my code

File: pkg/authorization/generated/clientset/typed/authorization/v1/authorization_client.go (hunk 1 above)
Comment (mfojtik) [inline pkg/authorization/generated/clientset/typed/authorization/v1/authorization_client.go]: ignore, shadowing...

**Comment by mfojtik** [comment]
@sttts @deads2k removed all weirdness and ready for review

**Comment by mfojtik** [comment]
i'll rebase on upstream https://github.com/openshift/origin/pull/15236 and pickup our patches

**Comment by mfojtik** [comment]
@deads2k @sttts another observation i made is that get on ImageStreamSecret actually return secrets and not "ImageStreamSecret"... I wonder if we want to teach client to handle 'unexpected' response...

Something like:

```
// +genclient:createInto=Secret
// for versioned:
// +genclient:createInto=v1/Secret
```

Another option is to change the API to return reasonable things but I guess it is too late for that.

**Comment by deads2k** [comment]
> I wonder if we want to teach client to handle 'unexpected' response...

I think we do for a different reason.  We need polymorphic endpoints like scale to be generated correctly.

Separate effort/pull though.

**Comment by mfojtik** [comment]
@sttts @deads2k so we have everything I need in kubernetes now to do this, however there is one problem. When I tried to pickup the change from kube, I found that the change is so big it will be better to do another kube rebase to that level... I don't think that is what we want to do now (or I want to commit to ;-)
Would it be reasonable to pick just generator changes for now as "<drop>" commit and use that to regenerate our clients? That seems less intrusive change to me than bumping kube just to pickup generator changes :-)

**Comment by sttts** [comment]
> Would it be reasonable to pick just generator changes for now as "" commit and use that to regenerate our clients?

Then on the next kube bump the unlucky person doing it has to remember to regenerate clients. Have you tried a bump? Are there any conflicts?

**Comment by mfojtik** [comment]
> Then on the next kube bump the unlucky person doing it has to remember to regenerate clients. Have you tried a bump? Are there any conflicts?

I will regenerate the clients so the next person will just drop that patch and everything should be working normally.

**Comment by mfojtik** [comment]
closing this, lets continue discussion here: https://github.com/openshift/origin/pull/15561
